dtfs-central-api/src/v1/controllers/portal/gef-facility/update-facility.controller.js

```diff
@@ -1,62 +1,81 @@
-const { MONGO_DB_COLLECTIONS } = require('@ukef/dtfs2-common');
+const { validateAuditDetails, generateAuditDatabaseRecordFromAuditDetails } = require('@ukef/dtfs2-common/change-stream');
+const { MONGO_DB_COLLECTIONS, InvalidAuditDetailsError, FacilityNotFoundError, InvalidDealIdError, InvalidFacilityIdError } = require('@ukef/dtfs2-common');
 const { ObjectId } = require('mongodb');
 const { mongoDbClient: db } = require('../../../../drivers/db-client');
 
-const updateFacility = async (id, updateBody) => {
-  if (!ObjectId.isValid(id)) {
-    return { status: 400, message: 'Invalid Facility Id' };
+const updateFacility = async ({ facilityId, facilityUpdate, auditDetails }) => {
+  if (!ObjectId.isValid(facilityId)) {
+    throw new InvalidFacilityIdError(facilityId);
   }
 
   try {
     const facilitiesCollection = await db.getCollection(MONGO_DB_COLLECTIONS.FACILITIES);
     const dealsCollection = await db.getCollection(MONGO_DB_COLLECTIONS.DEALS);
+    const auditRecord = generateAuditDatabaseRecordFromAuditDetails(auditDetails);
 
-    let updatedFacility;
+    const existingFacility = await facilitiesCollection.findOne({ _id: { $eq: ObjectId(facilityId) } });
 
-    const existingFacility = await facilitiesCollection.findOne({ _id: { $eq: ObjectId(id) } });
+    if (!existingFacility) {
+      throw new FacilityNotFoundError(facilityId);
+    }
 
-    if (existingFacility) {
-      const { dealId } = existingFacility;
+    const { dealId } = existingFacility;
 
-      if (dealId !== undefined && !ObjectId.isValid(dealId)) {
-        throw new Error('Invalid Deal Id');
-      }
+    if (dealId !== undefined && !ObjectId.isValid(dealId)) {
+      throw new InvalidDealIdError(dealId);
+    }
 
-      updatedFacility = await facilitiesCollection.findOneAndUpdate(
-        { _id: { $eq: ObjectId(id) } },
-        { $set: updateBody },
-        { returnNewDocument: true, returnDocument: 'after' },
-      );
+    const facilityUpdateWithAuditRecord = { ...facilityUpdate, auditRecord };
 
-      if (updatedFacility) {
-        // update facilitiesUpdated timestamp in the deal
-        const dealUpdateObj = { facilitiesUpdated: new Date().valueOf() };
+    const updatedFacility = await facilitiesCollection.findOneAndUpdate(
```

Comment:
If I remember right `findOneAndUpdate` returns a slightly weird & deprecated `ModifyResult` type in mongodb v4. I think the name of this should be something like `updateFacilityResponse`

Reply:
Using `updateOne` would be preferable but would need further refactoring